### PR TITLE
[CGP-411] Declarative Type System

### PR DESCRIPTION
**Pull request by psygnisfive**
Pursuant to [CGP 411](https://iohk.myjetbrains.com/youtrack/issue/CGP-411).

This branch replaces the main type system we currently have w/ one that uses a declarative equality for types only. In all the places where we rely on type normalization and normality, we explicit reduction and equality.

Of particular note, the judgment `Γ ⊢ M : S` is now `Γ ⊢ M : A` where the type can be an arbitrary type, thereby necessitating that every places we relied on its normality, we instead explicitly reduce. This means that in eliminators, where we would normally simply "match" on the synthesized type to verify that it has the right form (eg. `A → B` or `∀a. B`), we instead have to reduce it (eg. `C →ty* A → B`). Equality itself is only ever used to equate two synthesized types (i.e. in the rule for function application) because all other places we care about reduction, we only have the one synthesized type, not two types to compare.

This does suggest one way to efficiently implement this, however: normalize types only to WHNF unless equality is required, where instead we'd use full HNF.

Also note: The relation `→ty` is defined as relating two _types_, not a type and a normal type, which means that its transitive closure also relates two types. For many purposes that's fine, but we probably implicitly want somewhere to think instead of defining it as relating a type and a normal type, permitting the reflexive case only when the type is normal:

```
   A → A'    A' →* S
   ----------------
         A →* S

   -------
   S →* S
```

### CODE REVIEW

File: plutus-core-spec/figures/PlutusCoreTypeCorrectness.tex [hunk 1]
```diff
@@ -149,61 +149,66 @@
...
         \RightLabel{builtin}
-        \UnaryInfC{\(\hypJ{\Gamma}{\istermJ{\builtin{bn}{A_0, ..., A_m}{M_0, ..., M_n}}{\normalform{\subst{A_0, ..., A_m}{\alpha_0, ..., \alpha_m}{C}}}}\)}
+        \UnaryInfC{\(\hypJ{\Gamma}{\istermJ{\builtin{bn}{A_0 ... A_m}{M_0 ... M_n}}{\subst{A_0, ..., A_m}{\alpha_0, ..., \alpha_m}{C}}}\)}
     \end{prooftree}
 
     \begin{prooftree}
         \AxiomC{\(\hypJ{\Gamma}{\istypeJ{A}{\typeK{}}}\)}
         \RightLabel{error}
-        \UnaryInfC{\(\hypJ{\Gamma}{\istermJ{\error{A}}{\normalform{A}}}\)}
+        \UnaryInfC{\(\hypJ{\Gamma}{\istermJ{\error{A}}{A}}\)}
```
Comment (vmchale) [inline plutus-core-spec/figures/PlutusCoreTypeCorrectness.tex]: If we don't normalize here, type synthesis will not always produce normalized types.

**Comment by michaelpj** [comment]
Given that this is supposed to be the declarative version, I'd assumed that we would use `A = B` in all the places we currently have `A ->* B`. The former implies the latter, and I think it's equivalent when the RHS is a type that does not reduce.

File: plutus-core-spec/figures/PlutusCoreTypeCorrectness.tex (hunk 1 above)
Comment (michaelpj) [reply plutus-core-spec/figures/PlutusCoreTypeCorrectness.tex]: I think this is deliberate:

> Of particular note, the judgment Γ ⊢ M : S is now Γ ⊢ M : A where the type can be an arbitrary type

The algorithmic version will presumably have a judgement that produces normalized types.

**Comment by michaelpj** [comment]
You use an equivalence on kinds, but I don't think you define that anywhere.

**Comment by michaelpj** [comment]
The caption for the figure should probably be different now.

**Comment by psygnisfive** [comment]
@michaelpj 

> You use an equivalence on kinds, but I don't think you define that anywhere.

Mainly this was b/c we're now using `≡` on types to mean "syntactically identical", and `=` to mean "syntactically identical modulo beta reduction". It felt inconsistent to use `=` on kinds since they have not computation that we're incorporating, but I suppose most people will treat `=` as "the obvious equality for this sort of thing"? I'll change it never the less.

**Comment by psygnisfive** [comment]
@michaelpj 

> Given that this is supposed to be the declarative version, I'd assumed that we would use `A = B` in all the places we currently have `A ->* B`. The former implies the latter, and I think it's equivalent when the RHS is a type that does not reduce.

Reasonable, I suppose. It incorporates the computational definitions tho, given the way `=` is defined, but includes an unnecessary occurrence of `->*` hidden by the `=`. But it's equivalent, so that's not an issue formally.

File: plutus-core-spec/figures/PlutusCoreTypeCorrectness.tex (hunk 1 above)
Comment (psygnisfive) [reply plutus-core-spec/figures/PlutusCoreTypeCorrectness.tex]: I'm going to resolve this convo

**Comment by psygnisfive** [comment]
@michaelpj 

> The caption for the figure should probably be different now.

I think for this version the caption should be the same? But for the other two versions, it should be different, I agree.

**Pull request merged**